File: server/src/main/scala/edp/davinci/util/json/JsonUtils.scala
/*-
 * <<
 * Davinci
 * ==
 * Copyright (C) 2016 - 2017 EDP
 * ==
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * >>
 */





package edp.davinci.util.json

import org.json4s.JsonAST.JNothing
import org.json4s._
import org.json4s.jackson.JsonMethods._
import org.json4s.jackson.Serialization._

object JsonUtils extends JsonUtils

trait JsonUtils {
  implicit var json4sFormats: Formats = DefaultFormats

  // Case Class related
  def json2caseClass[T: Manifest](json: String): T = read[T](json)

  def caseClass2json[T <: AnyRef](obj: T): String = write[T](obj)

  def jsonCompact(json: String): String = compact(render(parse(json)))

  def jsonPretty(json: String): String = pretty(render(parse(json)))

  // JValue related
  def json2jValue(json: String): JValue = parse(json) //parseJsonString

  def jValue2json(jValue: JValue): String = compact(render(jValue)) //jsonToString

  def containsName(jValue: JValue, name: String): Boolean = jValue \ name != JNothing

  def getString(jValue: JValue, name: String): String = (jValue \ name).extract[String]

  def getInt(jValue: JValue, name: String): Int = (jValue \ name).extract[Int]

  def getLong(jValue: JValue, name: String): Long = (jValue \ name).extract[Long]

  def getShort(jValue: JValue, name: String): Short = (jValue \ name).extract[Short]

  def getBoolean(jValue: JValue, name: String): Boolean = (jValue \ name).extract[Boolean]

  def getList(jValue: JValue, name: String): List[JValue] = (jValue \ name).extract[List[JValue]]

  def getJValue(jValue: JValue, name: String): JValue = (jValue \ name).extract[JValue]
}

File: commons/src/main/java/edp/davinci/commons/util/JSONUtils.java
/*
 * <<
 *  Davinci
 *  ==
 *  Copyright (C) 2016 - 2020 EDP
 *  ==
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *        http://www.apache.org/licenses/LICENSE-2.0
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 *  >>
 *
 */

package edp.davinci.commons.util;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;

import com.fasterxml.jackson.core.JsonParser.Feature;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;

public class JSONUtils {
	
    private JSONUtils() {

    }
    
	private static void config(ObjectMapper mapper) {
		mapper.setDateFormat(new SimpleDateFormat("yyyy-MM-dd HH:mm:ss"));
		mapper.disable(SerializationFeature.FAIL_ON_EMPTY_BEANS);
		mapper.disable(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS);
		mapper.disable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES);
		mapper.enable(Feature.ALLOW_COMMENTS);
		mapper.enable(Feature.ALLOW_UNQUOTED_FIELD_NAMES);
		mapper.enable(Feature.ALLOW_SINGLE_QUOTES);
		mapper.enable(DeserializationFeature.USE_BIG_DECIMAL_FOR_FLOATS);
	}
    
	public static boolean validateObj(String jsonString) {

		jsonString = jsonString.trim();
		
		if (StringUtils.isEmpty(jsonString)) {
			return false;
		}

		try {
			ObjectMapper mapper = new ObjectMapper();
			mapper.readValue(jsonString, Map.class);
		} catch (Exception e) {
			return false;
		}

		return true;
	}
	
	public static boolean validateArray(String jsonString) {

		jsonString = jsonString.trim();

		if (StringUtils.isEmpty(jsonString)) {
			return false;
		}

		try {
			ObjectMapper mapper = new ObjectMapper();
			mapper.readValue(jsonString, mapper.getTypeFactory().constructParametricType(ArrayList.class, Map.class));
		} catch (Exception e) {
			return false;
		}

		return true;
	}
	
	public static String toString(Object obj) {
		return toString(obj, false);
	}

	public static String toString(Object obj, boolean prettyFormat) {
		ObjectMapper mapper = new ObjectMapper();
		config(mapper);
		mapper.enable(SerializationFeature.INDENT_OUTPUT);

		try {
			return mapper.writeValueAsString(obj);
		} catch (JsonProcessingException e) {
			e.printStackTrace();
		}
		
		return null;
	}
	
    public static <T> T toObject(String jsonString, Class<T> c) {

    	if (null == c || StringUtils.isEmpty(jsonString)) {
            return null;
        }
    	
    	ObjectMapper mapper = new ObjectMapper();
    	config(mapper);

        try {
			return mapper.readValue(jsonString, c);
		} catch (Exception e) {
			e.printStackTrace();
		}
        
        return null;
    }

    public static <T> List<T> toObjectArray(String jsonString, Class<T> c) {

        if (null == c || StringUtils.isEmpty(jsonString)) {
            return Collections.emptyList();
        }

        ObjectMapper mapper = new ObjectMapper();
    	config(mapper);

    	try {
			return mapper.readValue(jsonString, mapper.getTypeFactory().constructParametricType(ArrayList.class, c));
		} catch (Exception e) {
			e.printStackTrace();
		}
    	
    	return null;
    }
    
    public static Object convertJO2POJO(Object data) {

        Class<?> dCls = data.getClass();

        if (JSONObject.class.isAssignableFrom(dCls)) {

            Map<String, Object> m = new LinkedHashMap<String, Object>();

            JSONObject jod = (JSONObject) data;

            for (String key : jod.keySet()) {

                Object attr = jod.get(key);

                Object attrObj = convertJO2POJO(attr);

                m.put(key, attrObj);
            }

            return m;

        }
        else if (JSONArray.class.isAssignableFrom(dCls)) {

            List<Object> l = new ArrayList<Object>();

            JSONArray joa = (JSONArray) data;

            for (Object o : joa) {

                Object attrObj = convertJO2POJO(o);

                l.add(attrObj);
            }

            return l;

        }

        return data;
    }
}
